DOCS-11059 Snowflake Connector 1.0 (#1483)
mulesoft/docs-connectors: snowflake/1.0/modules/ROOT/pages/snowflake-connector-reference.adoc

** MINUTES
** HOURS
** DAYS |  Time unit for the *Query Timeout* field. |  SECONDS | 
| Fetch Size a| Number |  Indicates how many rows to fetch from the database when rows are read from a result set. This property is required when streaming is true, in which case a default value of 10 is used. |  | 
| Max Rows a| Number |  Sets the limit for the maximum number of rows that any result set object generated by the message processor can contain for the given number. If the limit is exceeded, the excess rows are silently dropped. |  | 
| SQL String a| String |  Text of the SQL query to execute. |  | x
| Parameter Types a| Array of <<ParameterType>> |  Allows you to optionally specify the type of one or more of the parameters in the query. If provided, you're not required to reference all of the parameters, but you can't reference a parameter that isn't present in the input values. |  | 
| Target Variable a| String |  Name of the variable that stores the operation's output. |  | 
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>

> * <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  | 

Reviewer: ```suggestion
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors. |  | 
```

| Target Variable a| String |  Name of the variable that stores the operation's output. |  | 
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors. |  | 
|===

==== Output
[%autowidth.spread]
|===
|Type |Array of Number
|===

=== For Configurations
* <<snowflake-config>> 

==== Throws
* SNOWFLAKE:RETRY_EXHAUSTED 
* SNOWFLAKE:CONNECTIVITY 
* SNOWFLAKE:NOT_FOUND 
* SNOWFLAKE:QUERY_EXECUTION 
* SNOWFLAKE:INVALID_CREDENTIALS 
* SNOWFLAKE:INVALID_INPUT 
* SNOWFLAKE:CANNOT_LOAD_DRIVER 
* SNOWFLAKE:UNKNOWN_SNOWFLAKE_ERROR 
* SNOWFLAKE:DATABASE 

[[copyIntoLocation]]
=== Copy Into Location
`<snowflake:copy-into-location>`

Unloads data from a table (or query) into one or more files.

==== Parameters
[%header,cols="20s,20a,35a,20a,5a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x
| Location a| One of:

* <<SnowflakeLocation>>

> * <<ExternalLocation>> |  Specifies the internal or external location where the files containing the data to be loaded are staged. |  | x

Reviewer: ```suggestion
* <<ExternalLocation>> |  Specifies the internal or external location where the files containing the data to load are staged. |  | x
```

| Table Query Source a| One of:

* <<table>>

> * <<query>> |  Specifies the table or query used as the source of the data to be unloaded. |  | x

Reviewer: ```suggestion
* <<query>> |  Specifies the table or query used as the source of the data to unload. |  | x
```

| File Format a| One of:

* <<DefinitionByNameForCopyIntoLocation>>
* <<DefinitionByTypeForCopyIntoLocation>> |  Specifies the format of the data files to load. |  | 
| Include Header a| Boolean |  Specifies whether to include the table column headings in the output files. |  false | 
| Transactional Action a| Enumeration, one of:

** ALWAYS_JOIN
** JOIN_IF_POSSIBLE
** NOT_SUPPORTED |  Type of joining action that operations can perform for transactions. |  JOIN_IF_POSSIBLE | 
| Overwrite a| Boolean |  Boolean that specifies whether the COPY command overwrites existing files with matching names in the location where the files are stored. |  false | 

> | Single a| Boolean |  Boolean that specifies whether to generate a single file or multiple files. If FALSE, a filename prefix must be included in the path. |  false | 

Reviewer: ```suggestion
| Single a| Boolean |  Boolean that specifies whether to generate a single file or multiple files. If `false`, a filename prefix must be included in the path. |  false | 
```

| Detailed Output a| Boolean |  Specifies whether to load data into columns in the target table that match corresponding columns represented in the data. |  false | 
| Target Variable a| String |  Name of the variable that stores the operation's output. |  | 
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  Retry strategy in case of connectivity errors. |  | 
|===

==== Output
[%autowidth.spread]
|===
|Type |Array of Object
|===

=== For Configurations
* <<snowflake-config>> 

==== Throws
* SNOWFLAKE:RETRY_EXHAUSTED 
* SNOWFLAKE:CONNECTIVITY 
* SNOWFLAKE:NOT_FOUND 
* SNOWFLAKE:QUERY_EXECUTION 
* SNOWFLAKE:INVALID_CREDENTIALS 
* SNOWFLAKE:INVALID_INPUT 
* SNOWFLAKE:CANNOT_LOAD_DRIVER 
* SNOWFLAKE:UNKNOWN_SNOWFLAKE_ERROR 
* SNOWFLAKE:DATABASE 

[[copyIntoTable]]
=== Copy Into Table
`<snowflake:copy-into-table>`

Loads data from staged files to an existing table. The files must already be staged in one of the following locations: 

> - Named internal stage (or table/user stage). Files can be staged using the PUT command. 

Reviewer: should these be bullets? (*)

Author: This format works the same as * I believe.

| Transactional Action a| Enumeration, one of:

** ALWAYS_JOIN
** JOIN_IF_POSSIBLE
** NOT_SUPPORTED |  Type of joining action that operations can perform for transactions. |  JOIN_IF_POSSIBLE | 
| On Error a| String |  Action to perform when an error is seen while loading data from a file. |  | 
| Size Limit a| Number | Maximum size in bytes of data to load with a given copy statement. |  | 
| Purge a| Boolean |  Specifies whether to remove the data files from the stage automatically after the data is loaded successfully. |  false | 
| Return Failed Only a| Boolean |  Specifies whether to return only files that have failed to load in the statement result. |  false | 
| Match By Column Name a| Enumeration, one of:

** CASE_SENSITIVE
** CASE_INSENSITIVE
** NONE |  Specifies whether to load data into columns in the target table that match corresponding columns represented in the data. |  NONE | 
| Enforce Length a| Boolean |  Specifies whether to truncate text strings that exceed the target column length. |  true | 
| Truncate Columns a| Boolean |  Specifies whether to truncate text strings that exceed the target column length. |  false | 

> | Force a| Boolean |  If true, loads all files, regardless of whether they've been loaded previously and have not changed since they were loaded. |  false | 

Reviewer: ```suggestion
| Force a| Boolean |  If `true`, loads all files, regardless of whether they've been loaded previously and have not changed since they were loaded. |  false | 
```

| Load Uncertain Files a| Boolean | Specifies to load files for which the load status is unknown. |  false | 
| Target Variable a| String |  Name of the variable that stores the operation's output. |  | 
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors. |  | 
|===

==== Output
[%autowidth.spread]
|===
|Type |Array of Object
|===

=== For Configurations
* <<snowflake-config>> 

==== Throws
* SNOWFLAKE:RETRY_EXHAUSTED 
* SNOWFLAKE:CONNECTIVITY 
* SNOWFLAKE:NOT_FOUND 
* SNOWFLAKE:QUERY_EXECUTION 
* SNOWFLAKE:INVALID_CREDENTIALS 
* SNOWFLAKE:INVALID_INPUT 
* SNOWFLAKE:CANNOT_LOAD_DRIVER 
* SNOWFLAKE:UNKNOWN_SNOWFLAKE_ERROR 
* SNOWFLAKE:DATABASE 

[[copyIntoTableWithTransformation]]
=== Copy Into Table With Transformation
`<snowflake:copy-into-table-with-transformation>`

Loads data from staged files to an existing table. The files must already be staged in one of the following locations: 

> - Named internal stage (or table/user stage). Files can be staged using the PUT command. 

Reviewer: bullets?

| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors. |  | 
|===

==== Output
[%autowidth.spread]
|===
|Type |Array of Object
|===

=== For Configurations
* <<snowflake-config>> 

==== Throws
* SNOWFLAKE:RETRY_EXHAUSTED 
* SNOWFLAKE:CONNECTIVITY 
* SNOWFLAKE:NOT_FOUND 
* SNOWFLAKE:QUERY_EXECUTION 
* SNOWFLAKE:INVALID_CREDENTIALS 
* SNOWFLAKE:INVALID_INPUT 
* SNOWFLAKE:CANNOT_LOAD_DRIVER 
* SNOWFLAKE:UNKNOWN_SNOWFLAKE_ERROR 
* SNOWFLAKE:DATABASE 

[[createPipe]]
=== Create Pipe
`<snowflake:create-pipe>`

Creates a new pipe in the system for defining the COPY INTO <table> statement used by Snowpipe to load data from an ingestion queue into tables.

==== Parameters
[%header,cols="20s,20a,35a,20a,5a"]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | Name of the configuration to use. | | x
| Pipe Name a| String |  Specifies the identifier for the pipe; must be unique for the schema in which the pipe is created. |  | x

> | Copy Statement a| String |  Specifies the statement used to load data from queued files into a Snowflake table. This statement serves as the text/definition for the pipe and is displayed in the SHOW PIPES output. |  | x

Reviewer: ```suggestion
| Copy Statement a| String |  Specifies the statement used to load data from queued files into a Snowflake table. This statement serves as the text or definition for the pipe and is displayed in the SHOW PIPES output. |  | x
```

| Transactional Action a| Enumeration, one of:

** ALWAYS_JOIN
** JOIN_IF_POSSIBLE
** NOT_SUPPORTED |  The type of joining action that operations can perform for transactions. |  JOIN_IF_POSSIBLE | 
| On Error a| String |  Action to perform when an error is seen while loading data from a file. |  | 
| Size Limit a| Number |  Maximum size in bytes of data to load with a given copy statement. |  | 
| Purge a| Boolean |  Specifies whether to remove the data files from the stage automatically after the data is loaded successfully. |  false | 
| Return Failed Only a| Boolean |  Specifies whether to return only files that have failed to load in the statement result. |  false | 
| Match By Column Name a| Enumeration, one of:

** CASE_SENSITIVE
** CASE_INSENSITIVE
** NONE |  Specifies whether to load data into columns in the target table that match corresponding columns represented in the data. |  NONE | 
| Enforce Length a| Boolean |  Specifies whether to truncate text strings that exceed the target column length. |  true | 
| Truncate Columns a| Boolean |  Specifies whether to truncate text strings that exceed the target column length. |  false | 

> | Force Load a| Boolean | If true, loads all files, regardless of whether they've been loaded previously and have not changed since they were loaded. |  false | 

Reviewer: ```suggestion
| Force Load a| Boolean | If `true`, loads all files, regardless of whether they've been loaded previously and have not changed since they were loaded. |  false | 
```

| Target Variable a| String |  Name of the variable that stores the operation's output. |  | 
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>

> * <<reconnect-forever>> |  A retry strategy in case of connectivity errors |  | 

Reviewer: ```suggestion
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors. |  | 
```

| Auto Generated Keys Column Names a| Array of String |  List of column names that indicates which auto-generated keys to make available for retrieval. |  | 
| Target Variable a| String |  Name of the variable that stores the operation's output. |  | 
| Target Value a| String |  Expression that evaluates the operation’s output. The outcome of the expression is stored in the *Target Variable* field. |  #[payload] | 
| Reconnection Strategy a| * <<reconnect>>
* <<reconnect-forever>> |  A retry strategy in case of connectivity errors. |  | 
|===

==== Output
[%autowidth.spread]
|===
|Type |<<StatementResult>>
|===

=== For Configurations
* <<snowflake-config>> 

==== Throws
* SNOWFLAKE:RETRY_EXHAUSTED 
* SNOWFLAKE:CONNECTIVITY 
* SNOWFLAKE:NOT_FOUND 
* SNOWFLAKE:QUERY_EXECUTION 
* SNOWFLAKE:INVALID_CREDENTIALS 
* SNOWFLAKE:INVALID_INPUT 
* SNOWFLAKE:CANNOT_LOAD_DRIVER 
* SNOWFLAKE:UNKNOWN_SNOWFLAKE_ERROR 
* SNOWFLAKE:DATABASE 

== Types

[[SnowflakeDbPoolingProfile]]
=== Snowflake Db Pooling Profile

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | Max Pool Size a| Number | Maximum number of connections a pool maintains at any given time | 5 | 

Reviewer: add periods for this type

| Min Pool Size a| Number | Minimum number of connections a pool maintains at any given time | 0 | 
| Acquire Increment a| Number | Determines how many connections at a time to try to acquire when the pool is exhausted | 1 | 
| Prepared Statement Cache Size a| Number | Determines how many statements are cached per pooled connection. Setting this to zero will disable statement caching | 5 | 
| Max Wait a| Number | The amount of time a client trying to obtain a connection waits for it to be acquired when the pool is
 exhausted. Zero (default) means wait indefinitely | 0 | 
| Max Wait Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS | A #maxWait. | SECONDS | 
| Max Idle Time a| Number | Determines how many seconds a Connection can remain pooled but unused before being discarded.
 Zero means idle connections never expire. | 0 | 
| Additional Properties a| Object |  |  | 
|===

[[Reconnection]]
=== Reconnection

Reconnection strategy to use. 

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | Fails Deployment a| Boolean | When the application is deployed, a connectivity test is performed on all connectors. If set to true, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  | 

Reviewer: ```suggestion
| Fails Deployment a| Boolean | When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  | 
```

* <<reconnect-forever>> | The reconnection strategy to use. |  | 
|===

[[reconnect]]
=== Reconnect

Configures a standard reconnection strategy, which specifies how often to reconnect and how many reconnection attempts the connector source or operation can make.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Frequency a| Number | How often in milliseconds to reconnect |  | 
| Count a| Number | How many reconnection attempts to make. |  | 
|===

[[reconnect-forever]]
=== Reconnect Forever

Configures a forever reconnection strategy by which the connector operation source or operation attempts to reconnect at a specified frequency for as long as the Mule app runs.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Frequency a| Number | How often in milliseconds to reconnect. |  | 
|===

[[ExpirationPolicy]]
=== Expiration Policy

Configures the minimum amount of time that a dynamic configuration instance can remain idle before Mule considers it eligible for expiration.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | Max Idle Time a| Number | A scalar time value for the maximum amount of time a dynamic configuration instance should be allowed to be idle before it's considered eligible for expiration |  | 

Reviewer: Add periods for this type

| Record Delimiter a| String | One or more single-byte or multibyte characters that separate records in an input file or unloaded file. |  | x
| Field Delimiter a| String | One or more single-byte or multibyte characters that separate records in an input file or unloaded file. | , | 
| File Extension a| String |  |  | 
| Skip Header Count a| Number |  | 0 | 
| Skip Blank Lines a| Boolean | Specifies whether to skip any blank lines encountered in data, instead of resulting in an end-of-record error. | false | 
| Date Format a| String |  | Auto | 
| Time Format a| String |  | Auto | 
| Timestamp Format a| String |  | Auto | 
| Binary Format a| Enumeration, one of:

** HEX
** BASE64
** UTF8 |  | HEX | 
| Escape Character a| String |  |  | 
| Escape Unenclosed Field Character a| String | Specifies a single character used as the escape character for unenclosed field values only. |  | 
| Trim Space a| Boolean |  | false | 
| Field Optionally Enclosed By a| String |  |  | 
| Null Ifs a| Array of String | Specifies a list of values from which Snowflake selects the first one to convert to from SQL NULL. |  | x

> | Errror On Column Count Mismatch a| Boolean | Specifies whether to generate a parsing error if the number of delimited columns (i.e. fields) in an input file does not match the number of columns in the corresponding table. | true | 

Reviewer: remove i.e.?

| Replace Invalid Characters a| Boolean | Specifies whether to replace invalid UTF-8 characters with a default Unicode character instead of resulting in an error. | false | 
| Validate UTF8 a| Boolean |  | true | 

> | Empty Field As Null a| Boolean | When set to false, allows empty strings to be loaded in tables without enclosing the field values in quotes. | true | 

Reviewer: ```suggestion
| Empty Field As Null a| Boolean | When set to `false`, allows empty strings to be loaded in tables without enclosing the field values in quotes. | true | 
```

| External Location Provider a| One of:

* <<MicrosoftAzure>>
* <<GoogleCloudStorage>>
* <<AmazonS3>> | External system to integrate with Snowflake |  | x
|===

[[InternalStageLocation]]
=== Internal Stage Location

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Stage Name a| String | Internal stage name. |  | x
|===

[[ExternalStageLocation]]
=== External Stage Location

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Stage Name a| String | External stage name. |  | x
| External Location Provider a| One of:

* <<MicrosoftAzure>>
* <<GoogleCloudStorage>>
* <<AmazonS3>> | External system to integrate with Snowflake |  | x
|===

[[MicrosoftAzure]]
=== Microsoft Azure

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Url a| String | URL for the Microsoft Azure. |  | x
| Microsoft Azure Storage Integration Credentials a| One of:

* <<MicrosoftAzureStorageIntegration>>
* <<MicrosoftAzureKeyCredentials>>
* <<MicrosoftAzureNoCredentials>> | Security credentials for connecting to Microsoft Azure. |  | x
| Encryption a| One of:

> * <<AzureCse>> | Required only for loading from/unloading into encrypted files; not required if storage location and files are unencrypted. |  | 

Reviewer: ```suggestion
* <<AzureCse>> | Required only for loading from or unloading into encrypted files; not required if storage location and files are unencrypted. |  | 
```

|===

[[GoogleCloudStorage]]
=== Google Cloud Storage

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Url a| String | URL for the Google Cloud Storage. |  | x
| Storage Integration Name a| String | Name of the storage integration used to delegate authentication responsibility for external cloud storage. |  | x
| Encryption a| One of:

> * <<GcsSseMks>> | Required only for loading from/unloading into encrypted files; not required if storage location and files are unencrypted. |  | 

Reviewer: global

|===

[[AmazonS3]]
=== Amazon S3

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Url a| String | URL for the Amazon S3. |  | x
| Amazon S3 Storage Integration Credentials a| One of:

* <<AmazonS3StorageIntegration>>
* <<AmazonS3KeyCredentials>>
* <<AmazonS3RoleCredentials>>
* <<AmazonS3NoCredentials>> | The security credentials for connecting to AWS and accessing the private/protected S3 bucket where the files to load/unload are staged. |  | x
| Encryption a| One of:

* <<AwsCse>>
* <<AwsSseS3>>
* <<AwsSseKms>> | Required only for loading from/unloading into encrypted files; not required if storage location and files are unencrypted. |  | 
|===

[[AwsCse]]

> === Aws Cse

Reviewer: ```suggestion
=== AWS CSE
```
